### DISCUSSION TITLE
How to add more configuration files to a web application?

### DISCUSSION THREAD

**Discussion by paulomorgado** [dotnet/aspnetcore#57304] [Q&A] [answered]
Is there an extension point to load more configuration file when using `WebApplication.CreateBuilder()` and `WebApplication.CreateBuilderSlim()`?

Is the only option to clear all configuration and load it all again?

**Comment by martincostello** [comment]
[`JsonConfigurationExtensions.AddJsonFile()`](https://learn.microsoft.com/dotnet/api/microsoft.extensions.configuration.jsonconfigurationextensions.addjsonfile)

For example:

```csharp
var builder = WebApplication.CreateBuilder();
builder.Configuration.AddJsonFile("whatever.json");
```

**Reply by paulomorgado** [reply]
@martincostello, that overrides anything that was supposed to have been overridden by environment variables and command-line arguments.

Configuration is roughly loaded in this order:

1. appsettings.json
2. appsettings{Environment}.json
3. Environment variables
4. Command line arguments

Configuration files need to be loaded before `3` and `4`.

**Reply by davidfowl** [reply, marked as answer]
You can mutate the list of sources, here's an example of inserting it first. 
```C#
var builder = WebApplication.CreateBuilder(args);

builder.Configuration.InsertBefore(c =>
{
    c.AddJsonFile("MyInitialSource.json", optional: true);
});

var app = builder.Build();

app.MapGet("/", () => "Hello World!");

app.Run();

public static class ConfigurationExtensions 
{
    public static ConfigurationManager InsertBefore(this ConfigurationManager config, Action<ConfigurationManager> action)
    {
        // Get all existing sources and make a copy
        var existing = config.Sources.ToArray();
        // Clear theses sources
        config.Sources.Clear();
        // Mutate configuration
        action(config);
        // Add the older sources back
        foreach (var s in existing)
        {
            config.Sources.Add(s);
        }
        return config;
    }
}
```

**Reply by paulomorgado** [reply]
> You can mutate the list of sources, here's an example of inserting it first. 
> 
> ```cs
> var builder = WebApplication.CreateBuilder(args);
> 
> builder.Configuration.InsertBefore(c =>
> {
>     c.AddJsonFile("MyInitialSource.json", optional: true);
> });
> 
> var app = builder.Build();
> 
> app.MapGet("/", () => "Hello World!");
> 
> app.Run();
> 
> public static class ConfigurationExtensions 
> {
>     public static ConfigurationManager InsertBefore(this ConfigurationManager config, Action<ConfigurationManager> action)
>     {
>         // Get all existing sources and make a copy
>         var existing = config.Sources.ToArray();
>         // Clear theses sources
>         config.Sources.Clear();
>         // Mutate configuration
>         action(config);
>         // Add the older sources back
>         foreach (var s in existing)
>         {
>             config.Sources.Add(s);
>         }
>         return config;
>     }
> }
> ```

Thanks.

I was just considering explicitly clearing the `Sources` and loading it all again.

Since `Sources` is an `IList<T>` I might be able to just `Insert` them.
